tests/unit/Request.ts

```diff
@@ -0,0 +1,391 @@
+		lang.mixin(this, {
+			_restore: function (Constructor: { new (arg: any): { restored: boolean } }) {
+				// use constructor based restoration
+				const restored = new Constructor(this);
+				restored.restored = true;
+				return restored;
+			},
+			createSubclass: null
+		}, args);
+	}
+
+	_restore(Constructor: { new (arg: any): { restored: boolean } }) {
+		// use constructor based restoration
+		const restored = new Constructor(this);
+		restored.restored = true;
+		return restored;
+	}
+
+	name: string;
+
+	describe() {
+		return 'name is ' + this.name;
+	}
+}
+
+
+class ConcreteRequest<T> extends Request<T> {
+	add(object: any, directives?: dstore.PutDirectives):  any {
+		throw new Error('This Method is abstract');
+	}
+
+	get(id: string | number): Promise<T> | void {
+		throw new Error('This Method is abstract');
+	}
+
+	put(object: T, directives?: dstore.PutDirectives):  Promise<T> {
+		return new Promise(function (resolve) {
+			resolve(object)
+		});
+	}
+
+	remove(id: any): Promise<T | void> {
+		throw new Error('This Method is abstract');
+	}
+}
+
+let store: ConcreteRequest<any>;
+
+const globalHeaders: { [ name: string ]: string } = {
+
+	'test-global-header-a': 'true',
+	'test-global-header-b': 'yes'
+};
+const requestHeaders: { [ name: string ]: string } = {
+	'test-local-header-a': 'true',
+	'test-local-header-b': 'yes',
+	'test-override': 'overridden'
+};
+
+function runCollectionTest<T>(collection: dstore.Collection<T>, rangeArgs: dstore.FetchRangeArgs | RequestOptions, expected?: RequestOptions) {
+	const expectedResults: { [ name: string ]: number | string }[] = [
+		{ id: 1, name: 'one' },
+		{ id: 2, name: 'two' }
+	];
+	mockRequestProvider.setResponseText(JSON.stringify(expectedResults));
+	let resultsPromise: dstore.FetchPromise<any>;
+	if (!expected) {
+		expected = <RequestOptions> rangeArgs;
+		resultsPromise = collection.fetch();
+	}
+	else {
+		const { start, end } = <dstore.FetchRangeArgs> rangeArgs;
+		mockRequestProvider.setResponseHeaders({
+			'Content-Range': start + '-' + end + '/' + expectedResults.length
+		});
+		resultsPromise = collection.fetchRange(<dstore.FetchRangeArgs> rangeArgs);
+	}
+	return resultsPromise.then(function (results) {
+		expected.headers && mockRequestProvider.assertRequestHeaders(expected.headers);
+		expected.query && mockRequestProvider.assertQuery(new UrlSearchParams(expected.query));
+
+		// We cannot just assert deepEqual with results and expectedResults
+		// because the store converts results into model instances with additional members.
+		assert.strictEqual(results.length, expectedResults.length);
+		for (let i = 0; i < results.length; ++i) {
+			const result: { [ name: string ]: number | string } = results[i];
+			const expectedResult: { [ name: string ]: number | string } = expectedResults[i];
+			for (let key in expectedResult) {
+				assert.strictEqual(result[key], expectedResult[key]);
+			}
+		}
+		if (resultsPromise.totalLength instanceof Promise) {
```

Comment:
It should always be there, but it could be a number. One case in which it would be a number is if QueryResults is passed an array. I believe that should no longer happen but I hadn't removed the logic while converting it. 

The second case is that If QueryOptions is passed in to QueryResults and has a totalLength property already specified, that value is used. I'm not sure when/if that is specified on the options passed to QueryResults or if it will ever be a number. If it isn't used or will always be a promise, then I could probably change this(and the type signature for totalLength) to treat it as if it will always be a promise.